system/mp/mpi_wrapper.cpp

```diff
@@ -0,0 +1,513 @@
+    {
+        return (send==_send) && (_rank==MPI_ANY_SOURCE || rank==_rank) && (_tag==MPI_ANY_TAG || tag==_tag) && (comm == _comm);
+    }
+
+    int rank;                       // source/destination rank of the processor
+    int tag;                        // MPI tag information
+    MPI::Request request;           // persistent request object to keep track of ongoing MPI call
+    MPI::Comm& comm;                // MPI communicator
+
+
+    CommData(bool _send, int _rank, int _tag, MPI::Comm& _comm):
+        send(_send), rank(_rank), tag(_tag), comm(_comm){}
+
+    CommData(int _rank, int _tag, MPI::Comm& _comm): rank(_rank), tag(_tag), comm(_comm), send(false){}
+
+    void notifyCancellation()
+    {
+        lock();
+        cancellationInProgress = true;
+        unlock();
+    }
+
+    bool lockFromCancellation()         //returns true only if currently some thread is not attempting to cancel
+    {
+        lock();
+        if (cancellationInProgress)
+        {
+            unlock();
+        } else
+        {
+            cancellationLock = true;
+        }
+        return !cancellationInProgress;
+    }
+
+    void releaseCancellationLock()
+    {
+        if (cancellationLock)
+        {
+            cancellationLock = false;
+            unlock();
+        }//TODO else clause: throw a meaningful error for invalid call for this function
+    }
+
+    ~CommData()
+    {
+
+    }
+};
+
+std::vector<CommData*> asyncCommData; // CommData list to manage while send/recv communication in progress
+CriticalSection commDataLock;         // A mutex lock for the index list above
+
+CommData* _popCommData(int index)
+{
+    _TF("_popCommData", index);
+    CommData* ret = NULL;
+    assertex(index>=0);
+    assertex(index < asyncCommData.size());
+    ret = asyncCommData[index];
+    asyncCommData.erase(asyncCommData.begin() + index);
+    return ret;
+}
+
+void addCommData(CommData *commData)
+{
+    _TF("addCommData", commData->rank, commData->tag);
+    CriticalBlock block(commDataLock);
+
+    //TODO Do a cleanup while we are at it
+//    int size = asyncCommData.size(); int completed; MPI_Status stat;
+//    for(int i=(size-1); i>=0 ; i--)
+//    {
+//        if (!(asyncCommData[i]->probingProgress))
+//        {
+//            completed = 0;
+//            assertex(asyncCommData[i]->request != NULL);
+//            _T("asyncCommData[i]->request="<<*(asyncCommData[i]->request)<<" mem_address="<<asyncCommData[i]->request);
+//            _T("send="<<asyncCommData[i]->isSend()<<" rank="<<asyncCommData[i]->rank<<" tag="<<asyncCommData[i]->tag);
+//            bool error = (MPI_Test(asyncCommData[i]->request, &completed, &stat)!= MPI_SUCCESS);
+//            if (completed || error) //unlikely an error would occur
+//            {
+//                delete _popCommData(i);
+//            }
+//        }
+//    }
+
+    asyncCommData.push_back(commData);
+}
+
+CommData* popCommData(int index)
```

Comment:
I don't think this method is used?
I don't think it would be safe if it were, because index obtained outside of the CS could be different after if it gets CS.

Reply:
yeah you are right. This was used in a previous implementation cycle was left behind when things changed. will remove it